src/Singer/Stream.hs

```diff
@@ -42,42 +43,44 @@ validTweet screenName t = not (retweeted t) && maybe True (== screenName) (inRep
 
 -- | Morphs Twitter source representation to a more strongly typed model
 toTweet :: (Monad m) => S.Stream (S.Of TweetSrc) m r -> S.Stream (S.Of Tweet) m r
-toTweet = S.catMaybes . S.map fromTweetSrc
+toTweet = S.mapMaybe fromTweetSrc
 
 -- | A base filter that only removes answers and retweets
 baseFilter :: (Monad m) => Text -> S.Stream (S.Of TweetSrc) m r -> S.Stream (S.Of TweetSrc) m r
 baseFilter screenName = S.filter (validTweet screenName)
 
-foundParent :: (Monad m) => S.Stream (S.Of Tweet) m () -> Maybe Tweet -> Tweet -> S.Stream (S.Of Thread) m ()
-foundParent src child' parent =
-  let withChild = parent { child = child' }
-      asThread = maybe (toThread' src Nothing) S.yield . mkThread $ withChild
-  in maybe asThread (const $ toThread' src (Just withChild)) (replyTo parent)
+foundParent :: Maybe Tweet -> Tweet -> Maybe Tweet
+foundParent child' parent = pure $ parent { child = child' }
 
-didNotFindParent :: (Monad m) => S.Stream (S.Of Tweet) m () -> Maybe Tweet -> Tweet -> S.Stream (S.Of Thread) m ()
-didNotFindParent src child' parent =
-  let onIsReply = toThread' src (pure parent)
-      onIsNotReply = toThread' src Nothing
-      startNewThread = maybe onIsNotReply (const onIsReply) (replyTo parent)
-      continueCurrentThread = toThread' src child'
-  in maybe startNewThread (const continueCurrentThread) child'
+didNotFindParent :: Maybe Tweet -> Tweet -> Maybe Tweet
+didNotFindParent child' parent = let
+  ongoingThread = void . replyTo
+  verifiedChild = do
+    t <- child'
+    ongoingThread t
+    child'
+  in verifiedChild <|> pure parent
+
+extractThread :: Maybe Tweet -> Maybe Thread
+extractThread tweet = do
+  t <- tweet
+  completedThread t
+  hasMultipleTweets t
```

Comment:
I'm not sure I understand this line, would you explain a bit what you're doing here ?

Reply:
I (vim actually) use britanny for formatting, so I basically did nothing but saving the files.
If you're talking about extract thread, the idea is to emit a thread if and only if we finished an actual thread, so we check if the threads is completed (no replyTo in the last tweet) and if it is actually a thread (it has replies), and we use the do-notation of the maybe monad for a clearer (maybe not so clearer, as you had to ask) syntax.

Reply:
Ah - right, I think I have the idea - should child be nothing, the do block would stop here. Fair enough !

Reply:
Yes, that's why there's a `void` there: I don't care about the child content, I just want to know that there's one. It ends up with a `Maybe ()` just like `guard`.